core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -1227,6 +1227,16 @@ private void substituteSubQuery(Blackboard bb, SubQuery subQuery) {
         // Otherwise, let convertExists translate
         // values list into an inline table for the
         // reference to Q below.
+      } else if (query instanceof SqlSelect) {
+        // During JOIN ON condition processing, bb.root is
+        // initially null but subquery expansion requires
+        // a non-null bb.root. For IN/NOT IN subqueries
+        // processed via convertExists, we use the join's
+        // left input as bb.root. Exception: in UPDATE queries,
+        // bb.inputs remains null and subquery rewriting is skipped.
+        if (bb.inputs != null && bb.inputs.size() == 2) {
```

Comment:
is there a way to assert here that this is indeed inside a JOIN?
I am worried about other patterns matching.

Reply:
It's not easy to obviously demonstrate the problem in a quidem test, but I did enter my changes when debugging.

Reply:
This request is not about quidem tests.
The comments says "we use the join's left input".
How do you know that's from a join?
Can you check that?

Reply:
Currently, only when there is a subquery in the join on condition, `bb.root` will be null and `query` will be of `SqlSelect` type. I wrote `bb.inputs.size() == 2` in the code to confirm that `bb.inputs` does have the two i`nputs` to join.

Reply:
there are other operators with 2 inputs

Reply:
I understand that `join` is the only method that can have subqueries and multiple `inputs`. I didn't expect other cases. The `substituteSubQuery` method also involves subqueries. I don't know if I understand it correctly.

Reply:
why is the left input to the join the correct root?
what if that is itself a join?

Reply:
If the subquery is in `where`, the `inputs` may be a `join`. If the subquery is in the `on` condition of a `join`, the `inputs` will be `left` and `right`. The `root` used in the [register](https://github.com/apache/calcite/pull/4354/files#diff-5d7c774cd39d94b119eedcf3d361e72406c7864e916736c533bbe8f893ca72bcR1336 ) method is `left`.

Reply:
Yes, but the docs say that the root cannot be a join.
I am wondering about the case when the left input of the join is a join.
Maybe you can at least `assert !(root instanceof Join)` after the assignment?
Moreover, directly assigning to root is strange, I have never seen this pattern in Calcite.
I would have expected a use of setRoot.
It is true that there is another example in this file, and that code was written by @julianhyde.

Reply:
Oh, I can use setRoot instead.

Reply:
Frankly, I am not sure this fix is correct for all possible queries (because I don't really know how the blackboard works), but it is an improvement, since some correct queries which used to be rejected are now correctly compiled, so I will approve.

Reply:
> Currently, only when there is a subquery in the join on condition, bb.root will be null and query will be of SqlSelect type. I wrote bb.inputs.size() == 2 in the code to confirm that bb.inputs does have the two inputs to join

Is it possible to use `bb.scope` to check if it's a join ? The current statement is probably correct for now, but it might be difficult to debug if any issues come up later

Reply:
@dssysolyatin Thanks for your suggestion, that might be a good idea.